src/Extensions/Logging/Internal/EventSource/OmexLogEventSender.cs

```diff
@@ -0,0 +1,182 @@
+					break;
+				case LogLevel.Information:
+					m_eventSource.LogInfoServiceMessage(applicationName, serviceName, machineId, buildVersion, s_processName, partitionId, replicaId, activityId, traceIdAsString, "Info", category, tagId, tagName, threadId, message);
+					break;
+				case LogLevel.Warning:
+					m_eventSource.LogWarningServiceMessage(applicationName, serviceName, machineId, buildVersion, s_processName, partitionId, replicaId, activityId, traceIdAsString, "Warning", category, tagId, tagName, threadId, message);
+					break;
+				case LogLevel.Error:
+				case LogLevel.Critical:
+					m_eventSource.LogErrorServiceMessage(applicationName, serviceName, machineId, buildVersion, s_processName, partitionId, replicaId, activityId, traceIdAsString, "Error", category, tagId, eventId.Name, threadId, message);
+					break;
+				default:
+					throw new ArgumentException("Unknown EventLevel: " + level);
+			}
+		}
+
+
+		public bool IsEnabled(LogLevel level) =>
+		level switch
+		{
+			LogLevel.None => false,
+			_ => m_eventSource.IsEnabled()
+		};
+
+
+		public bool IsReplayableMessage(LogLevel logLevel) =>
+			logLevel switch
+			{
+				LogLevel.Trace => true,
+				LogLevel.Debug => true,
+				_ => false
+			};
+
+
+		public void ReplayLogs(Activity activity)
+		{
+			if (activity is ReplayableActivity replayableActivity)
+			{
+				foreach (LogMessageInformation log in replayableActivity.GetLogEvents())
+				{
+					LogMessage(activity.Id, activity.TraceId, log.Category, LogLevel.Information, log.EventId, log.ThreadId, log.Message);
+				}
+			}
+		}
+
+
+		private readonly OmexLogEventSource m_eventSource;
+		private readonly IServiceContext m_serviceContext;
+		private readonly IMachineInformation m_machineInformation;
+		private static readonly string s_processName;
+
+
+		/// <summary>
+		/// Get the Tag id as a string
+		/// </summary>
+		/// <param name="tagId">tag id</param>
+		/// <returns>the tag as string</returns>
+		/// <remarks>
+		/// In terms of the conversion from integer tag value to equivalent string reprsentation, the following scheme is used:
+		/// 1. If the integer tag &lt;= 0x0000FFFF, treat the tag as special tag called numeric only tag.
+		/// Hence the string representation is direct conversion i.e. tag id 6700 == 6700
+		/// 2. Else, if it's an alphanumeric tag, there are 2 different schemes to pack those. viz. 4 letter and 5 letter representations.
+		/// 2.1 four letter tags are converted by transforming each byte into it's equivalent ASCII. e.g. 0x61626364 => abcd
+		/// 2.2 five letter tags are converted by transforming lower 30 bits of the integer value into the symbol space a-z,0-9.
+		/// The conversion is done by treating each group of 6 bits as an index into the symbol space a,b,c,d, ... z, 0, 1, 2, ....9
+		/// eg. 0x000101D0 = 00 000000 000000 010000 000111 010000 2 = aaqhq
+		/// </remarks>
+		private static string TagIdAsString(int tagId) // Convert should be done more efficient
```

Comment:
Is the comment here supposed to be a TODO?

Reply:
It's not todo because I'm not sure how to optimize it. I just don't like creating three arrays for tag convert and wanted to think about it after completing with other parts of the system.

Reply:
I'll remove comment to avoid confusion if I'll figure out how to do it better I'll create an issue or PR